File: apps/python/lib/chapter12/__main__.py
import pandas as pd
from sklearn.linear_model import Lasso
from sklearn.metrics import r2_score

from lib.chapter12.boston_features import (
    FEATURES,
    OUTLIER_THRESHOLD,
    TARGET,
    prepare_boston,
    remove_outliers,
)
from lib.chapter12.regularization import (
    best_experiment,
    fit_ridge,
    predict,
    run_ridge_experiments,
    zero_coefficient_names,
)
from lib.dataset import data_dir

TEST_SIZE = 0.3
VALIDATION_SIZE = 0.3
SEED = 0
ALPHAS = [0.0, 0.1, 1.0, 10.0, 100.0]
LASSO_ALPHA = 0.5


def main() -> None:
    csv_file = data_dir() / "Boston.csv"
    df = pd.read_csv(csv_file)
    kept = remove_outliers(df, FEATURES + [TARGET], OUTLIER_THRESHOLD)
    dataset = prepare_boston(csv_file, TEST_SIZE, VALIDATION_SIZE, SEED)
    print(f"データ件数: {len(kept)}（外れ値 {len(df) - len(kept)} 件を除外）")
    print(
        f"訓練データ: {len(dataset.t_train)} 件, "
        f"検証データ: {len(dataset.t_valid)} 件, "
        f"テストデータ: {len(dataset.t_test)} 件"
    )
    print(f"特徴量: {', '.join(dataset.feature_names)}")

    experiments = run_ridge_experiments(
        dataset.x_train, dataset.t_train, dataset.x_valid, dataset.t_valid, ALPHAS
    )
    print("alpha  訓練 R²  検証 R²  係数の絶対値の合計")
    for e in experiments:
        print(
            f"{e.alpha:>5}  {e.train_score:.4f}  {e.validation_score:.4f}  "
            f"{e.coef_abs_sum:.3f}"
        )
    best = best_experiment(experiments)
    print(f"検証データで選んだ alpha: {best.alpha}")

    linear = fit_ridge(dataset.x_train, dataset.t_train, alpha=0.0)
    ridge = fit_ridge(dataset.x_train, dataset.t_train, alpha=best.alpha)
    linear_score = r2_score(dataset.t_test, predict(linear, dataset.x_test))
    ridge_score = r2_score(dataset.t_test, predict(ridge, dataset.x_test))
    print(
        f"テストデータの決定係数: 線形回帰 {linear_score:.4f}, "
        f"リッジ回帰 {ridge_score:.4f}"
    )

    lasso = Lasso(alpha=LASSO_ALPHA).fit(dataset.x_train, dataset.t_train)
    zeros = ", ".join(zero_coefficient_names(lasso.coef_, dataset.feature_names))
    print(f"ラッソ回帰（alpha={LASSO_ALPHA}）で係数が 0 になった特徴量: {zeros}")


if __name__ == "__main__":
    main()

File: apps/node/src/chapter12/main.ts
import { join } from "node:path";
import { dataDir } from "../dataset.ts";
import {
  type BostonDataset,
  FEATURES,
  OUTLIER_THRESHOLD,
  TARGET,
  loadBoston,
  prepareBoston,
  removeOutliers,
} from "./boston-features.ts";
import { fitLasso } from "./lasso.ts";
import {
  type Experiment,
  bestExperiment,
  fitRidge,
  predict,
  r2Score,
  runRidgeExperiments,
  zeroCoefficientNames,
} from "./regularization.ts";

const TEST_SIZE = 0.3;
const VALIDATION_SIZE = 0.3;
const SEED = 0;
const ALPHAS = [0, 0.1, 1, 10, 100];
const LASSO_LAMBDA = 0.05;
const SEEDS_TO_COMPARE = [0, 1, 2, 3, 4];

interface Comparison {
  experiments: readonly Experiment[];
  best: Experiment;
  linearScore: number;
  ridgeScore: number;
}

/** 検証データで alpha を選び、線形回帰と選んだリッジ回帰をテストデータで 1 回だけ評価する */
function compareOnTestData(dataset: BostonDataset): Comparison {
  const experiments = runRidgeExperiments(dataset, ALPHAS);
  const best = bestExperiment(experiments);
  const score = (alpha: number) =>
    r2Score(
      dataset.tTest,
      predict(fitRidge(dataset.xTrain, dataset.tTrain, alpha), dataset.xTest),
    );
  return {
    experiments,
    best,
    linearScore: score(0),
    ridgeScore: score(best.alpha),
  };
}

export function main(print: (line: string) => void = console.log): void {
  const csvFile = join(dataDir(), "Boston.csv");
  const rows = loadBoston(csvFile);
  const kept = removeOutliers(rows, [...FEATURES, TARGET], OUTLIER_THRESHOLD);
  const dataset = prepareBoston(csvFile, TEST_SIZE, VALIDATION_SIZE, SEED);
  print(
    `データ件数: ${kept.length}（外れ値 ${rows.length - kept.length} 件を除外）`,
  );
  print(
    `訓練データ: ${dataset.tTrain.length} 件, 検証データ: ${dataset.tValid.length} 件, テストデータ: ${dataset.tTest.length} 件`,
  );
  print(`特徴量: ${dataset.featureNames.join(", ")}`);

  const { experiments, best, linearScore, ridgeScore } =
    compareOnTestData(dataset);
  print("alpha\t訓練 R²\t検証 R²\t係数の絶対値の合計");
  for (const e of experiments) {
    print(
      `${e.alpha}\t${e.trainScore.toFixed(4)}\t${e.validationScore.toFixed(4)}\t${e.coefficientAbsSum.toFixed(3)}`,
    );
  }
  print(`検証データで選んだ alpha: ${best.alpha}`);
  print(
    `テストデータの決定係数: 線形回帰 ${linearScore.toFixed(4)}, リッジ回帰 ${ridgeScore.toFixed(4)}`,
  );

  const lasso = fitLasso(dataset.xTrain, dataset.tTrain, LASSO_LAMBDA);
  const zeros = zeroCoefficientNames(lasso.coefficients, dataset.featureNames);
  print(
    `ラッソ回帰（lambda=${LASSO_LAMBDA}）で係数が 0 になった特徴量: ${zeros.join(", ")}`,
  );

  print("");
  print("シード\t選んだ alpha\t線形回帰\tリッジ回帰");
  for (const seed of SEEDS_TO_COMPARE) {
    const c = compareOnTestData(
      prepareBoston(csvFile, TEST_SIZE, VALIDATION_SIZE, seed),
    );
    print(
      `${seed}\t${c.best.alpha}\t${c.linearScore.toFixed(4)}\t${c.ridgeScore.toFixed(4)}`,
    );
  }
}

// node src/chapter12/main.ts で直接実行したときだけ main を呼ぶ
if (import.meta.filename === process.argv[1]) {
  main();
}
